internal/pkg/agent/application/gateway/fleet/fleet_gateway.go

```diff
@@ -155,38 +179,156 @@ func (f *fleetGateway) Run(ctx context.Context) error {
-			resp, err := f.doExecute(ctx, backoff)
-			if err != nil {
+func (f *fleetGateway) performCancellableCheckin(ctx context.Context, initialState state.State, stateUpdates <-chan state.State, backoff backoff.Backoff, dbouncer debouncer) checkinResult {
+	checkinID, _ := uuid.NewV4()
+	checkinCtx, cancelCheckinCtxFunc := context.WithCancel(ctx)
+	checkinCtx = context.WithValue(checkinCtx, CheckinIDKey, checkinID.String())
+	f.log.Debugf("starting checkin %q", checkinID.String())
+	defer func() {
+		// make sure we don't leak contexts whatever happens
+		cancelCheckinCtxFunc()
+	}()
+
+	// Execute the checkin call asynchronously. For any errors returned by the fleet-server API
+	// the function will retry to communicate with fleet-server with an exponential delay and some
+	// jitter to help better distribute the load from a fleet of agents.
+	resCheckinChan := f.doExecuteAsync(checkinCtx, backoff, initialState)
+
+	// this variable will hold the latest state we got from coordinator
+	var updatedState *state.State
+	var debounceElapsed bool
+
+	for {
+		select {
+		case <-ctx.Done():
+			return checkinResult{err: ctx.Err()}
+		case checkinResult := <-resCheckinChan:
+			return checkinResult
+		case <-dbouncer.Elapsed():
+			debounceElapsed = true
+			if updatedState != nil && !reflect.DeepEqual(updatedState, &initialState) {
+				// we detected that the updated State is different from the one of the current checkin, cancel and return the needNewCheckinError
+				f.log.Debugf("Detected an updated state at the end of the debounce, cancelling ongoing checkin %q", checkinID)
+				cancelCheckinTimeoutCtx, cancelCancelCheckin := context.WithTimeout(ctx, f.cancelTimeout)
+				defer cancelCancelCheckin()
+				f.cancelCheckin(cancelCheckinTimeoutCtx, checkinID.String(), cancelCheckinCtxFunc, resCheckinChan)
+				return checkinResult{err: &needNewCheckinError{newState: *updatedState}}
+			}
+			f.log.Debug("No new state detected at the end of the debounce")
+		case newState := <-stateUpdates:
+			if !debounceElapsed {
+				f.log.Debugf(
+					"Updated state (Agent state: %q) received but current checkin %s not cancelled because we are still within debounce",
+					newState.State,
+					checkinID,
+				)
+				updatedState = &newState
 				continue
 			}
 
-			actions := make([]fleetapi.Action, len(resp.Actions))
-			copy(actions, resp.Actions)
-			if len(actions) > 0 {
-				f.actionCh <- actions
-			}
+			f.log.Infof(
+				"Received updated state (Agent state: %q) when checkin is ongoing past configured debounce. Cancelling previous checkin %q and starting a new one.",
+				newState.State,
+				checkinID.String(),
+			)
+			cancelCheckinTimeoutCtx, cancelCancelCheckin := context.WithTimeout(ctx, f.cancelTimeout)
+			defer cancelCancelCheckin()
+			f.cancelCheckin(cancelCheckinTimeoutCtx, checkinID.String(), cancelCheckinCtxFunc, resCheckinChan)
+			return checkinResult{err: &needNewCheckinError{newState: newState}}
```

Comment:
So, if I understand correctly, you would like to have something like a small delay before triggering a cancel in case the state goes back to what was previously... That is gonna be interesting to shoehorn in the current code structure...

I could start refactoring the debounce right away instead of waiting for #2504 if you like the proposed solution

Reply:
Yes.

I am good with the solution proposed in #2504 for the debounce as long as its a wrapper over the subscription and not something that changes the subscription code and messaging (some things need the information instantly without debounce).

Reply:
The idea here is to have 2 debouncers that have an input channel, an output channel and an internal state for the last value received.
If we create a small pipe composed of a one-off 5 mins debouncer and a resettable 30 seconds debouncer we should be fine and the subscription code has no changes

Reply:
I like the implementation idea.

I would prefer to see it in this PR, because I worry that we will drop it and not get it done. I would rather have it out of the gate instead of waiting.

Reply:
After our discussion yesterday, I added a debouncer which behaves like before but it allows for cleaner code in 781ccb7aee7fe9b7b15dadb870b122f4ae6f46ca

I also tried to define a debouncer which will return a bit of history of the states received during the debounce period in 42aae248551848f7fbfae3ca3d90bc3ae5ae06fd so that we can consolidate a state by discarding transient ones (we need to define what amounts as transient) but even this will not be a perfect solution (it will heavily depend on the heuristics we use).

I am going to try and integrate the AccumulatorDebouncer in the fleet gateway but I am starting to feel that we are trying to solve a different problem with this and we need to discuss what amounts as "good enough" for this PR.

Reply:
As agreed with @cmacknz, @blakerouse, the behaviour implemented with a simple debounce is no worse than what we already have.
Debouncing of transient agent states will not be included in this PR but it will be a followup issue.